torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -813,38 +813,52 @@ at::IntArrayRef strides_or_error(const Tensor & input, c10::string_view const &
       "Please either use a strided tensor or set requires_grad=False for '",
       input_name, "'");
     if (input.is_mkldnn()) return IntArrayRef({});
+    if (input.is_sparse_csr()) return IntArrayRef({});
     return input.strides();
   } else {
     return IntArrayRef({});
   }
 }
 
-Tensor mm_mat1_backward(const Tensor & grad, const Tensor & mat2, at::IntArrayRef mat1_sizes, at::IntArrayRef mat1_strides, const Scalar & alpha) {
-  // if input was column-major, return grad as column-order for efficiency
-  if (mat1_strides[0] == 1 && mat1_strides[1] == mat1_sizes[0]) {
-    return maybe_multiply(mat2.conj().mm(grad.t()).t(), alpha.conj());
-  } else {
-    return maybe_multiply(grad.mm(mat2.t().conj()), alpha.conj());
-  }
-}
-
-Tensor mm_mat2_backward(const Tensor & grad, const Tensor & mat1, IntArrayRef sizes, IntArrayRef strides, const Scalar & alpha) {
-  // if input was column-major, return grad as column-order for efficiency
-  if (strides[0] == 1 && strides[1] == sizes[0]) {
-    if (mat1.is_sparse()) {
-      // Since mm(dense, sparse) doesn't exist,
-      // pass a transposed output matrix to the underlying "addmm"
-      // function directly.
-      int64_t out_rows = mat1.size(1);
-      int64_t out_cols = grad.size(1);
-      Tensor t = at::zeros({}, grad.options()).expand({out_rows, out_cols}, true);
-      Tensor r = at::empty({out_cols, out_rows}, grad.options()).t();
-      at::addmm_out(r, t, mat1.t(), grad, alpha, 1);
-      return r;
+Tensor mm_mat1_backward(const Tensor& grad, const Tensor& mat2, at::IntArrayRef mat1_sizes, at::IntArrayRef mat1_strides, c10::Layout mat1_layout, const Scalar& alpha) {
+  if (grad.layout() == c10::kStrided && mat2.layout() == c10::kStrided && mat1_layout == c10::kStrided) {
+    // if input was column-major, return grad as column-order for efficiency
+    if (mat1_strides[0] == 1 && mat1_strides[1] == mat1_sizes[0]) {
+      return maybe_multiply(mat2.conj().mm(grad.t()).t(), alpha.conj());
     }
-    return maybe_multiply(grad.t().mm(mat1.conj()).t(), alpha.conj());
-  } else {
-    return maybe_multiply(mat1.t().conj().mm(grad), alpha.conj());
+  }
+
+  auto result = maybe_multiply(grad.mm(mat2.t().conj()), alpha.conj());
+
+  switch (mat1_layout) {
+    case c10::kStrided:
+      return result.to_dense();
+    case c10::kSparseCsr:
+      return result.to_sparse_csr();
```

Comment:
I'm not sure what's the point of doing this? we know that the gradient of mat1 will be sparse only if mat2 is sparse. So is there a benefit at doing this?
If the goal is only to silence the autograd engine, then I think we can make sure the layout check skips are extended for csr to match what is done for coo.

Reply:
It's not allowed to have a gradient with a layout different from the primal tensor's layout. I think layout checks are not skipped for coo, it's also not allowed to have a dense gradient for sparse_coo tensor.

Reply:
@IvanYashchuk - to be more precise we allow (Dense, Dense) and (*, Sparse) according to https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/engine.cpp#L713-L724 - We could extend this for CSR.

Here is the snippet
```
    if (grad.layout() != metadata.layout()) {
       // TODO: Currently we only support (*, Sparse) combination for (tensor.layout(), tensor.grad.layout())
       // In future, there will be an oppportunity to support more combinations of layouts if they are composable
       // (example., operations like addition etc., are well defined between tensors of different layouts.),
       // as well as all parts of autograd like AccumulateGrad correctly handle this.
       if (!grad.is_sparse()) {
        std::stringstream ss;
        ss << "invalid gradient at index " << i << " - expected layout ";
        ss << metadata.layout() << " but got " << grad.layout();
        AT_ERROR(format_error(ss.str()));
       }
    }
```

Reply:
Alright, now it's allowed to have (SparseCsr, Strided) combination for `(tensor.layout(), tensor.grad.layout())`. And I removed converting the layout of the `result`.